File: EPI/9.1/swift/9.1.swift
public struct Stack<T: Comparable> {
    private var elements = [T]()
    private var maximums = [(element: T, count: Int)]()

    public var isEmpty: Bool {
        return elements.isEmpty
    }

    public var max: T {
        return maximums.last!.element
    }

    public mutating func push(element: T) {
        elements.append(element)

        let lastMax = maximums.last
        if lastMax == nil || element > lastMax!.element {
            maximums.append((element, 1))
        } else if lastMax != nil && element == lastMax!.element {
            maximums[maximums.count - 1].count += 1
        }
    }

    public mutating func pop() -> T {
        let popped = elements.removeLast()

        if let lastMax = maximums.last where lastMax.element == popped {
            if lastMax.count > 1 {
                maximums[maximums.count - 1].count = 0
            } else {
                maximums.removeLast()
            }
        }

        return popped
    }
}

File: EPI/Stacks and Queues/Implement a stack with max API/typescript/9.1.ts
export default class Stack<T> {
    private elements: T[] = [];
    private maxIndexes: number[] = [];

    constructor(private compare: (a: T, b: T) => number) {
    }

    private maxIndex(): number {
        return this.maxIndexes.length === 0 ? -1 : this.maxIndexes[this.maxIndexes.length - 1];
    }

    max(): T | undefined {
        const maxIndex = this.maxIndex();
        return maxIndex < 0 ? undefined : this.elements[maxIndex];
    }

    push(element: T): void {
        if (this.maxIndexes.length === 0 || this.compare(this.max()!, element) < 0) {
            this.maxIndexes.push(this.elements.length);
        }

        this.elements.push(element);
    }

    pop(): T | undefined {
        if (this.maxIndex() === this.elements.length - 1) {
            this.maxIndexes.pop();
        }

        return this.elements.pop();
    }
}
